File: cpp_pract/merge_sorted.cpp
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    void insertAndBump(vector<int>& nums1, int i, int elem) {

        int tmp;
        for (auto start = nums1.begin() + i; start != nums1.end(); start++) {
            if (start == nums1.begin() + i) {
                tmp = *start;
                *start = elem;
            } else {
                int newTemp = *start;
                *start = tmp;
                tmp = newTemp;
            }
        }
    }
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        if (n == 0) {
            return;
        }
        if (m == 0) {
            for(int i = 0; i < n; i++) {
                nums1[i] = nums2[i];
            }
            return;
        }
        int n1Pos = 0;
        int n2Pos = 0;
        int insertionPos = 0;
        while (insertionPos < (m + n)) {
            if (*(nums1.begin() + n1Pos) > *(nums2.begin() + n2Pos)) {
                insertAndBump(nums1, insertionPos, *(nums2.begin() + n2Pos));
                n2Pos ++;
            } else {
                if (n1Pos < m) {
                    n1Pos++;
                } else {
                    insertAndBump(nums1, insertionPos, *(nums2.begin() +  n2Pos));
                    n2Pos++;
                }
            }
            insertionPos++;
        }
    }
};

int main() {
    Solution s = Solution();
    vector<int> testA = vector<int>{2,5,6,7,0};
    vector<int> testB = vector<int>{1};
    // s.insertAndBump(testA, 2, 9);
    s.merge(testA, 4, testB, 1);
    for (auto aIt = testA.begin(); aIt != testA.end(); aIt++) {
        cout << *aIt <<", ";
    }
}

File: merge_sorted.py
class Solution(object):
    def insertAndBump(self,nums1, i, elem):
        tmp = 0
        for niter in range(i, len(nums1)):
            if (niter == i):
                tmp = nums1[niter]
                nums1[niter] = elem
            else:
                newTemp = nums1[niter]
                nums1[niter] = tmp
                tmp = newTemp
    def merge(self,nums1, m, nums2, n):
        if (n == 0):
            return
        if (m == 0):
            for i in range(0,n):
                nums1[i] = nums2[i]
            return
        n1Pos = 0
        n2Pos = 0
        insertionPos = 0
        insertedFromN2 = 0
        while insertionPos < (m + n):
            if (n2Pos < len(nums2) and n1Pos < len(nums1) and nums1[n1Pos] > nums2[n2Pos]):
                nums1.insert(insertionPos, nums2[n2Pos])
                nums1.pop(len(nums1)-1)
                n2Pos += 1
                n1Pos = insertionPos + 1
                insertedFromN2 += 1
            else:
                if (n1Pos < (insertedFromN2 + m)):
                    n1Pos += 1
                elif n2Pos < n:
                    nums1.insert(insertionPos, nums2[n2Pos])
                    nums1.pop(len(nums1)-1)
                    n2Pos += 1
            insertionPos += 1

s = Solution()
res = [1,2,3,0,0,0]
s.merge(res, 3, [2,5,6], 3)
print(res)